modules/storages/app/common/storages/peripherals/storage_interaction/authentication_strategies/nextcloud_strategies.rb

```diff
@@ -37,10 +37,24 @@ module NextcloudStrategies
             ::Storages::Peripherals::StorageInteraction::AuthenticationStrategies::BasicAuth.strategy
           end
 
-          UserBound = ->(user:) do
-            ::Storages::Peripherals::StorageInteraction::AuthenticationStrategies::OAuthUserToken
-              .strategy
-              .with_user(user)
+          UserBound = ->(user:, storage:) do
+            # README: This check does not match for user that are only hijacked by an IDP based login.
+            token_exchange_preferred = storage.audience.present? && user.authentication_provider.present?
+
+            if token_exchange_preferred
+              ::Storages::Peripherals::StorageInteraction::AuthenticationStrategies::SsoUserToken
+                .strategy
+                .with_user(user)
+            elsif storage.oauth_configuration.present?
+              # README: This allows for potentially optional configuration of OAuth authentication method.
+              ::Storages::Peripherals::StorageInteraction::AuthenticationStrategies::OAuthUserToken
+                .strategy
+                .with_user(user)
+            else
+              # README: If for a user - storage combination no possible authentication method is found, the
+              # noop strategy will result in non authorized requests.
+              ::Storages::Peripherals::StorageInteraction::AuthenticationStrategies::Noop.strategy
```

Comment:
Will this result in us making actual requests that will then be rejected?

IIRC we talked about how we should avoid sending out requests that we know will fail eventually.

Would it be better to cause an error/exception here that can be caught before a request goes out?